File: server/python/src/metaobjects/migrate/expected_views.py
"""Build view bodies from projection metadata (origins).

Port of TS expected-views.ts / Java ViewBodyBuilder. v1 supports passthrough
(no @via) + aggregate (@agg/@of/@via). Deferred to v2 (returns None →
caller drops the view from the snapshot): passthrough WITH @via, collection,
multi-base.

Identifiers quoted throughout so mixed-case columns (e.g. "programId")
round-trip through PG's case-fold.
"""
from __future__ import annotations

from ..meta.meta_root import MetaRoot
from ..meta.core.object.meta_object import MetaObject
from ..meta.core.identity.meta_identity import MetaIdentity
from ..meta.core.identity import identity_constants as ic
from ..meta.core.field.meta_field import MetaField
from ..meta.core.field import field_constants as fc
from ..meta.core.relationship.meta_relationship import MetaRelationship
from ..meta.persistence.source.meta_source import MetaSource
from ..meta.persistence.source import source_constants as sc
from ..meta.persistence.origin.meta_origin import MetaOrigin
from ..meta.persistence.origin import origin_constants as oc
from .types import ViewDescriptor


def build_expected_views(root: MetaRoot, entity_by_name: dict[str, MetaObject]) -> list[ViewDescriptor]:
    out: list[ViewDescriptor] = []
    for child in root.own_children():
        if not isinstance(child, MetaObject):
            continue
        primary = _primary_read_only_source(child)
        if primary is None or primary.table_name() is None:
            continue
        # Skip write-through (has both writable AND read-only sources).
        if _primary_writable_source(child) is not None:
            continue
        body = _build_view_body(child, entity_by_name)
        if body is None:
            continue
        out.append(ViewDescriptor(name=primary.table_name(), schema=primary.schema(), sql=body))
    return out


def _build_view_body(projection: MetaObject, entity_by_name: dict[str, MetaObject]) -> str | None:
    """Compute the SELECT body, or None when unresolvable."""
    cols: list[str] = []
    base_entity: str | None = None
    T = "t"  # alias inside correlated subqueries

    for f in projection.fields():
        origin = _first_origin(f)
        if origin is None:
            return None
        field_col = _column_of(f)

        # passthrough (no @via) → plain column off the single base
        if origin.sub_type == oc.ORIGIN_SUBTYPE_PASSTHROUGH and origin.attr(oc.ORIGIN_ATTR_VIA) is None:
            from_attr = origin.attr(oc.ORIGIN_ATTR_FROM)
            if not isinstance(from_attr, str) or "." not in from_attr:
                return None
            ent, fname = from_attr.split(".", 1)
            bare = _strip_pkg(ent)
            if base_entity is None:
                base_entity = bare
            elif base_entity != bare:
                return None  # multi-base — deferred
            src = entity_by_name.get(base_entity)
            src_col = _column_for(src, fname)
            cols.append(f'  "{src_col}" AS "{field_col}"')
        # aggregate (@agg/@of/@via) → correlated subquery over a to-many
        elif origin.sub_type == oc.ORIGIN_SUBTYPE_AGGREGATE:
            agg = origin.attr(oc.ORIGIN_ATTR_AGG)
            of_attr = origin.attr(oc.ORIGIN_ATTR_OF)
            via = origin.attr(oc.ORIGIN_ATTR_VIA)
            if not (isinstance(agg, str) and isinstance(of_attr, str) and isinstance(via, str)
                    and "." in of_attr and "." in via):
                return None
            base_ent, rel_name = via.split(".", 1)
            bare = _strip_pkg(base_ent)
            if base_entity is None:
                base_entity = bare
            elif base_entity != bare:
                return None
            fk = _resolve_to_many_fk(entity_by_name, base_entity, rel_name)
            if fk is None:
                return None
            of_col = _column_for(fk.target, of_attr.split(".", 1)[1])
            base_table = _base_table(entity_by_name, base_entity)
            cols.append(
                f'  (SELECT {agg}({T}."{of_col}") FROM "{fk.target_table}" {T} '
                f'WHERE {T}."{fk.fk_col}" = "{base_table}"."{fk.parent_col}") AS "{field_col}"'
            )
        else:
            # passthrough WITH via OR collection — deferred
            return None

    if base_entity is None or not cols:
        return None
    return "SELECT\n" + ",\n".join(cols) + f'\nFROM "{_base_table(entity_by_name, base_entity)}"'


# ----------------------------------------------------------------------------
# To-many FK resolution
# ----------------------------------------------------------------------------


class _ToManyFk:
    """(target_entity, target_table, fk-col-on-target, parent-pk-col-on-base)."""
    __slots__ = ("target", "target_table", "fk_col", "parent_col")

    def __init__(self, target: MetaObject, target_table: str, fk_col: str, parent_col: str) -> None:
        self.target = target
        self.target_table = target_table
        self.fk_col = fk_col
        self.parent_col = parent_col


def _resolve_to_many_fk(
    entity_by_name: dict[str, MetaObject], base_entity_name: str, rel_name: str,
) -> _ToManyFk | None:
    base_obj = entity_by_name.get(base_entity_name)
    if base_obj is None:
        return None
    rel = next(
        (r for r in base_obj.children() if isinstance(r, MetaRelationship) and r.name == rel_name),
        None,
    )
    if rel is None or rel.object_ref() is None:
        return None
    target = entity_by_name.get(_strip_pkg(rel.object_ref()))
    if target is None:
        return None
    target_primary = _primary_writable_source(target)
    if target_primary is None or target_primary.table_name() is None:
        return None

    # Find the target's ReferenceIdentity pointing back at base_entity_name.
    fk_ref: MetaIdentity | None = None
    for c in target.children():
        if not isinstance(c, MetaIdentity) or c.sub_type != ic.IDENTITY_SUBTYPE_REFERENCE:
            continue
        ref = c.attr(ic.IDENTITY_REFERENCE_ATTR_REFERENCES)
        if isinstance(ref, str) and _strip_pkg(ref.split(".", 1)[0]) == base_entity_name:
            fk_ref = c
            break
    if fk_ref is None:
        return None
    fk_fields = _identity_fields(fk_ref)
    if not fk_fields:
        return None
    fk_col = _column_for(target, fk_fields[0])

    # Parent column on base — explicit target field else PK first field
    ref_attr = fk_ref.attr(ic.IDENTITY_REFERENCE_ATTR_REFERENCES)
    if isinstance(ref_attr, str) and "." in ref_attr:
        parent_field = ref_attr.split(".", 1)[1]
    else:
        pi = base_obj.primary_identity()
        parent_field = _identity_fields(pi)[0] if pi is not None and _identity_fields(pi) else "id"
    parent_col = _column_for(base_obj, parent_field)
    return _ToManyFk(target=target, target_table=target_primary.table_name(), fk_col=fk_col, parent_col=parent_col)


# ----------------------------------------------------------------------------
# Helpers (shared with expected_schema.py — duplicated to avoid a cycle)
# ----------------------------------------------------------------------------


def _first_origin(f: MetaField) -> MetaOrigin | None:
    for c in f.own_children():
        if isinstance(c, MetaOrigin):
            return c
    return None


def _identity_fields(identity: MetaIdentity) -> tuple[str, ...]:
    raw = identity.attr(ic.IDENTITY_ATTR_FIELDS)
    if raw is None:
        return ()
    if isinstance(raw, str):
        return (raw,)
    if isinstance(raw, (list, tuple)):
        return tuple(str(x) for x in raw)
    return ()


def _column_of(field: MetaField) -> str:
    col = field.attr(fc.FIELD_ATTR_COLUMN)
    return col if isinstance(col, str) and col else field.name


def _column_for(entity: MetaObject | None, field_name: str) -> str:
    if entity is None:
        return field_name
    f = entity.find_field(field_name)
    return _column_of(f) if f is not None else field_name


def _primary_writable_source(entity: MetaObject) -> MetaSource | None:
    for child in entity.own_children():
        if isinstance(child, MetaSource) and child.role() == sc.SOURCE_ROLE_PRIMARY and child.is_writable():
            return child
    return None


def _primary_read_only_source(entity: MetaObject) -> MetaSource | None:
    for child in entity.own_children():
        if isinstance(child, MetaSource) and child.role() == sc.SOURCE_ROLE_PRIMARY and child.is_read_only():
            return child
    return None


def _base_table(entity_by_name: dict[str, MetaObject], entity_name: str) -> str:
    e = entity_by_name.get(entity_name)
    if e is None:
        return entity_name
    p = _primary_writable_source(e)
    return p.table_name() if (p and p.table_name()) else entity_name


def _strip_pkg(name: str) -> str:
    idx = name.rfind("::")
    return name if idx == -1 else name[idx + 2 :]

File: server/typescript/packages/migrate-ts/src/expected-views.ts
// expected-views.ts — derive ViewDescriptor[] from projection metadata.
//
// Ports the shape of csharp/MetaObjects.Codegen/Schema/PostgresSchema.cs's
// CreateView. v1 supports:
//
//   * passthrough origin (no @via)  → plain column from the single base entity
//   * aggregate origin (@agg/@of/@via) → correlated subquery over a to-many
//
// Deferred to v2 (returns a "-- TODO" view that the runner skips applying):
//
//   * passthrough origin WITH @via  → to-one correlated subquery
//   * collection origin              → json_agg over a to-many
//   * multi-base projections         → blocked
//
// Identifiers are quoted throughout so mixed-case column names (e.g.
// "programId") survive PG's case-folding pass.

import type { MetaObject, MetaRoot } from "@metaobjectsdev/metadata";
import {
  type ColumnNamingStrategy,
  MetaPassthroughOrigin,
  MetaAggregateOrigin,
  MetaCollectionOrigin,
  MetaSource,
  SOURCE_ROLE_PRIMARY,
  TYPE_OBJECT,
  TYPE_ORIGIN,
  resolveColumnName,
  resolveTableName,
  stripPackage,
} from "@metaobjectsdev/metadata";
import type { ViewDescriptor } from "./types.js";

export function buildExpectedViews(root: MetaRoot, strategy: ColumnNamingStrategy): ViewDescriptor[] {
  const out: ViewDescriptor[] = [];
  for (const child of root.ownChildren()) {
    if (child.type !== TYPE_OBJECT) continue;
    const proj = child as MetaObject;
    if (proj.isAbstract) continue;
    if (!isReadOnlyProjection(proj)) continue;
    const view = buildView(proj, root, strategy);
    if (view !== null) out.push(view);
  }
  return out;
}

function isReadOnlyProjection(entity: MetaObject): boolean {
  const sources = entity.ownChildren().filter((c): c is MetaSource => c instanceof MetaSource);
  if (sources.length === 0) return false;
  const hasReadOnly = sources.some((s) => s.isReadOnly());
  const hasWritable = sources.some((s) => !s.isReadOnly());
  return hasReadOnly && !hasWritable;
}

function buildView(
  projection: MetaObject, root: MetaRoot, strategy: ColumnNamingStrategy,
): ViewDescriptor | null {
  const primarySource = projection.ownChildren().find(
    (c): c is MetaSource => c instanceof MetaSource && c.role === SOURCE_ROLE_PRIMARY,
  );
  // Use physicalName (FR-016 kind-aware), NOT tableName — a view-kind source names
  // itself via @view, and tableName reads ONLY the @table slot (undefined for views),
  // which previously bailed every @view projection here before any column was built.
  if (primarySource === undefined) return null;
  const viewName = primarySource.physicalName;

  const cols: string[] = [];
  let baseEntity: string | undefined;
  let blocked: string | undefined;
  const T = "t"; // target alias in correlated subqueries

  const baseTable = () => {
    if (baseEntity === undefined) return undefined;
    const found = root.findObject(baseEntity);
    return found ? resolveTableName(found) : baseEntity;
  };

  for (const f of projection.fields()) {
    const origin = f.ownChildren().find((c) => c.type === TYPE_ORIGIN);
    const fieldCol = resolveColumnName(f, strategy);

    // FR-024 base-link: a projection field that uses `extends: Base.field` (instead of
    // an explicit origin) is an implicit passthrough — its source IS the extended field.
    // That linkage lives in EXTENDED metadata (superRef / resolveSuper()), which
    // f.ownChildren() cannot see, so without this an extends-based field (typically the
    // PK that declares the projection's base) falls through to "no resolvable origin" and
    // bails the whole view — taking the sibling origin.passthrough renames down with it.
    if (origin === undefined && f.superRef !== undefined && f.superRef.includes(".")) {
      const [ent, field] = splitDot(f.superRef);
      const bare = stripPackage(ent);
      baseEntity ??= bare;
      if (bare !== baseEntity) { blocked = "passthrough from multiple base entities"; break; }
      const srcEntity = root.findObject(baseEntity);
      const srcCol = resolveColumnByName(srcEntity, field, strategy);
      cols.push(`  "${srcCol}" AS "${fieldCol}"`);
      continue;
    }

    if (origin instanceof MetaPassthroughOrigin && origin.via === undefined &&
        origin.from !== undefined && origin.from.includes(".")) {
      const [ent, field] = splitDot(origin.from);
      const bare = stripPackage(ent);
      baseEntity ??= bare;
      if (bare !== baseEntity) { blocked = "passthrough from multiple base entities"; break; }
      const srcEntity = root.findObject(baseEntity);
      const srcCol = resolveColumnByName(srcEntity, field, strategy);
      cols.push(`  "${srcCol}" AS "${fieldCol}"`);
    } else if (origin instanceof MetaAggregateOrigin && origin.agg !== undefined &&
               origin.of !== undefined && origin.via !== undefined &&
               origin.of.includes(".") && origin.via.includes(".")) {
      const [baseEnt, relName] = splitDot(origin.via);
      const bareBase = stripPackage(baseEnt);
      baseEntity ??= bareBase;
      if (bareBase !== baseEntity) { blocked = "aggregate over a different base entity"; break; }
      const fk = resolveToManyFk(root, baseEntity, relName, strategy);
      if (fk === null) {
        blocked = `unresolved to-many FK for @via "${origin.via}" (target needs an identity.reference back to ${baseEntity})`;
        break;
      }
      const ofCol = resolveColumnByName(fk.target, splitDot(origin.of)[1], strategy);
      cols.push(
        `  (SELECT ${origin.agg}(${T}."${ofCol}") ` +
        `FROM "${fk.targetTable}" ${T} ` +
        `WHERE ${T}."${fk.fkCol}" = "${baseTable()}"."${fk.parentCol}") AS "${fieldCol}"`,
      );
    } else if (origin instanceof MetaPassthroughOrigin || origin instanceof MetaCollectionOrigin) {
      blocked = `field "${f.name}" uses an origin shape not yet supported by TS migrate-ts (passthrough-via / collection — deferred)`;
      break;
    } else {
      blocked = `field "${f.name}" has no resolvable origin`;
      break;
    }
  }

  if (blocked !== undefined || baseEntity === undefined || cols.length === 0) {
    // Caller decides what to do with a null result. For v1 the runner just
    // omits it from expected.views — diff sees no expected view, no
    // create-view change, and an actual leftover view (if any) gets dropped.
    return null;
  }

  const sql = `SELECT\n${cols.join(",\n")}\nFROM "${baseTable()}"`;
  const view: ViewDescriptor = { name: viewName, sql };
  return view;
}

function splitDot(s: string): [string, string] {
  const i = s.indexOf(".");
  return [s.slice(0, i), s.slice(i + 1)];
}

function resolveColumnByName(
  owner: MetaObject | undefined, fieldName: string, strategy: ColumnNamingStrategy,
): string {
  if (owner === undefined) return fieldName;
  const field = owner.fields().find((f) => f.name === fieldName);
  return field ? resolveColumnName(field, strategy) : fieldName;
}

interface ToManyFk {
  target: MetaObject;
  targetTable: string;
  /** Column on the target entity holding the FK back to the base. */
  fkCol: string;
  /** Column on the base entity the FK references (usually the base's PK). */
  parentCol: string;
}

function resolveToManyFk(
  root: MetaRoot, baseEntityName: string, relName: string, strategy: ColumnNamingStrategy,
): ToManyFk | null {
  const baseObj = root.findObject(baseEntityName);
  if (baseObj === undefined) return null;
  const rel = baseObj.relationships().find((r) => r.name === relName);
  if (rel === undefined || rel.objectRef === undefined) return null;
  const target = root.findObject(stripPackage(rel.objectRef));
  if (target === undefined) return null;

  const fkRef = target.referenceIdentities().find(
    (r) => r.targetEntity !== undefined && stripPackage(r.targetEntity) === baseEntityName,
  );
  if (fkRef === undefined) return null;
  const fkFields = fkRef.fields;
  if (fkFields.length === 0) return null;
  const fkCol = resolveColumnByName(target, fkFields[0]!, strategy);

  const parentFieldName = fkRef.targetFields.length > 0
    ? fkRef.targetFields[0]!
    : baseObj.primaryIdentity()?.fields[0];
  if (parentFieldName === undefined) return null;
  const parentCol = resolveColumnByName(baseObj, parentFieldName, strategy);

  return { target, targetTable: resolveTableName(target), fkCol, parentCol };
}

